src/osd/PGLog.cc

```diff
@@ -453,7 +454,6 @@ void PGLog::merge_log(pg_info_t &oinfo, pg_log_t&& olog, pg_shard_t fromosd,
     for (auto &&oe: divergent) {
       dout(10) << "merge_log divergent " << oe << dendl;
     }
-    log.roll_forward_to(log.head, rollbacker);
```

Comment:
I think PR can be simplified and made more unit-testable by only doing a `roll_forward_to()` if `new_entries` is not empty, which is essentially what is used for doing `log->add()` and setting `should_rollforward`.